src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/LanguageServices/WorkspaceFactory.cs

```diff
@@ -0,0 +1,147 @@
+    }
+
+    public Workspace Create(
+        IActiveConfigurationSubscriptionSource source,
+        ProjectConfigurationSlice slice,
+        JoinableTaskFactory joinableTaskFactory,
+        Guid projectGuid,
+        CancellationToken cancellationToken)
+    {
+        UpdateHandlers updateHandlers = new(_handlerFactories);
+
+        Workspace workspace = new(
+            slice,
+            _unconfiguredProject,
+            projectGuid,
+            updateHandlers,
+            _logger,
+            _activeWorkspaceProjectContextTracker,
+            _commandLineParsers,
+            _dataProgressTrackerService,
+            _workspaceProjectContextFactory,
+            _faultHandlerService,
+            joinableTaskFactory,
+            _threadingService.JoinableTaskContext,
+            unloadCancellationToken: cancellationToken);
+
+        ITargetBlock<IProjectVersionedValue<WorkspaceUpdate>> actionBlock
+            = DataflowBlockFactory.CreateActionBlock<IProjectVersionedValue<WorkspaceUpdate>>(
+                update => joinableTaskFactory.RunAsync(async () => await workspace.OnWorkspaceUpdateAsync(update)),
+                _unconfiguredProject,
+                ProjectFaultSeverity.LimitedFunctionality,
+                nameFormat: "Workspace update handler {0}");
+
+        #region Evaluation data
+
+        var evaluationTransformBlock
+            = DataflowBlockSlim.CreateTransformBlock<IProjectVersionedValue<(ConfiguredProject ConfiguredProject, IProjectSubscriptionUpdate EvaluationRuleUpdate, IProjectSubscriptionUpdate SourceItemsUpdate)>, IProjectVersionedValue<WorkspaceUpdate>>
+                (update => update.Derive(WorkspaceUpdate.FromEvaluation));
+
+        workspace.ChainDisposal(new DisposableBag
+        {
+            ProjectDataSources.SyncLinkTo(
+                source.ActiveConfiguredProjectSource.SourceBlock.SyncLinkOptions(),
+                source.ProjectRuleSource.SourceBlock.SyncLinkOptions(DataflowOption.WithRuleNames(updateHandlers.EvaluationRules)),
+                source.SourceItemsRuleSource.SourceBlock.SyncLinkOptions(),
+                target: evaluationTransformBlock,
+                linkOptions: DataflowOption.PropagateCompletion,
+                cancellationToken: cancellationToken),
+
+            evaluationTransformBlock.LinkTo(actionBlock, DataflowOption.PropagateCompletion),
+
+            ProjectDataSources.JoinUpstreamDataSources(joinableTaskFactory, _projectService.Services.FaultHandler, source.ProjectRuleSource, source.SourceItemsRuleSource)
+        });
+
+        #endregion
+
+        #region Build data subscriptions
+
+        // A block that provides the command line arguments for a configured project.
+        // We will link this to the slice's active configured project.
+        // NOTE this will go away once CPS exposes a way to access snapshot data while preserving order (being worked on for 17.4)
```

Comment:
Link for tracking?

Reply:
CPS has already implemented this. I have a local branch that'll address this. I don't want to make this PR any more complex though.